arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -291,6 +294,153 @@ void DatabaseManagerThread::run() {
+  std::string testFilePath = FileUtils::buildFilename(
+      databasePathFeature.directory(), "TestFileIOHeartbeat");
+  std::string testFileContent = "This is just an I/O test.\n";
+
+  LOG_TOPIC("66665", DEBUG, Logger::ENGINES) << "IOHeartbeatThread: running...";
+
+  while (true) {
+    try {  // protect thread against any exceptions
+      if (isStopping()) {
+        // done
+        break;
+      }
+
+      LOG_TOPIC("66659", DEBUG, Logger::ENGINES)
+          << "IOHeartbeat: testing to write/read/remove " << testFilePath;
+      // We simply write a file and sync it to disk in the database
+      // directory and then read it and then delete it again:
+      auto start1 = std::chrono::steady_clock::now();
+      bool trouble = false;
+      try {
+        FileUtils::spit(testFilePath, testFileContent, true);
+      } catch (std::exception const& exc) {
+        ++_failures;
+        LOG_TOPIC("66663", INFO, Logger::ENGINES)
+            << "IOHeartbeat: exception when writing test file: " << exc.what();
+        trouble = true;
+      }
+      auto finish = std::chrono::steady_clock::now();
+      std::chrono::duration<double> dur = finish - start1;
+      bool delayed = dur > std::chrono::seconds(1);
+      if (trouble || delayed) {
+        if (delayed) {
+          ++_delays;
+        }
+        LOG_TOPIC("66662", INFO, Logger::ENGINES)
+            << "IOHeartbeat: trying to write test file took "
+            << std::chrono::duration_cast<std::chrono::microseconds>(dur)
+                   .count()
+            << " milliseconds.";
+      }
+
+      // Read the file if we can reasonably assume it is there:
+      if (!trouble) {
+        auto start = std::chrono::steady_clock::now();
+        try {
+          std::string content = FileUtils::slurp(testFilePath);
+          if (content != testFileContent) {
+            LOG_TOPIC("66660", INFO, Logger::ENGINES)
+                << "IOHeartbeat: read content of test file was not as "
+                   "expected, found:'"
+                << content << "', expected: '" << testFileContent << "'";
+            trouble = true;
+            ++_failures;
+          }
+        } catch (std::exception const& exc) {
+          ++_failures;
+          LOG_TOPIC("66661", INFO, Logger::ENGINES)
+              << "IOHeartbeat: exception when reading test file: "
+              << exc.what();
+          trouble = true;
+        }
+        auto finish = std::chrono::steady_clock::now();
+        std::chrono::duration<double> dur = finish - start;
+        bool delayed = dur > std::chrono::seconds(1);
+        if (trouble || delayed) {
+          if (delayed) {
+            ++_delays;
+          }
+          LOG_TOPIC("66669", INFO, Logger::ENGINES)
+              << "IOHeartbeat: trying to read test file took "
+              << std::chrono::duration_cast<std::chrono::microseconds>(dur)
+                     .count()
+              << " milliseconds.";
```

Comment:
The value reported is in microseconds, but the log messages says "milliseconds".

Reply:
LOL. Will fix.